File: Backend/sporty/src/main/java/tn/esprit/sporty/Service/ProductService.java
package tn.esprit.sporty.Service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import tn.esprit.sporty.Entity.*;
import tn.esprit.sporty.Repository.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RequiredArgsConstructor
@Service
public class ProductService implements IProductService {
    private final IProductRepository productRepository;
    private final IShopRepository shopRepository;
    private final IWhishlistRepository whishlistRepository;
    private final IReviewRepository reviewRepository;
    private final IPhoneNumberRepository phoneNumberRepository;



    @Override
    public Product addProduct(Product product) {
        return productRepository.save(product);
    }

    @Override
    public Product retrieveProduct(Long idProduct) {
        return productRepository.findById(idProduct).orElse(null);
    }

    @Override
    public List<Product> retrieveAll() {
        return (List<Product>) productRepository.findAll();
    }

    @Override
    public void deleteProduct(Long idProduct) {
        productRepository.deleteById(idProduct);
    }

    @Override
    public Product updateProduct(Product product) {
        return productRepository.save(product);
    }
@Override
public Product assignProductToShop(Long productId, Long shopId) {
    Optional<Product> productOptional = productRepository.findById(productId);
    Optional<Shop> shopOptional = shopRepository.findById(shopId);

    if (productOptional.isPresent() && shopOptional.isPresent()) {
        Product product = productOptional.get();
        Shop shop = shopOptional.get();

        product.setShop(shop); // Assign the shop to the product
        return productRepository.save(product); // Save the updated product
    } else {
        throw new RuntimeException("Product or Shop not found!");
    }
}
    @Override
    public List<Product> getProductsByPriceRange(float minPrice,float maxPrice) {
        return productRepository.findByPriceBetween(minPrice, maxPrice);
    }
    @Override
    public List<Product> searchProductsByName(String productName) {
        return productRepository.findByProductNameContainingIgnoreCase(productName);
    }
    @Override
    public Map<String, Double> getCategoryDistribution() {
        List<Product> products = productRepository.findAll();  // Fetch all products

        // Map to store the count of products per category
        Map<String, Long> categoryCount = new HashMap<>();
        // Total number of products
        long totalProducts = products.size();

        // Count products per category
        for (Product product : products) {
            String category = product.getFootballProductsCategory().name();
            categoryCount.put(category, categoryCount.getOrDefault(category, 0L) + 1);
        }

        // Calculate percentage for each category
        Map<String, Double> categoryPercentage = new HashMap<>();
        for (Map.Entry<String, Long> entry : categoryCount.entrySet()) {
            categoryPercentage.put(entry.getKey(), (entry.getValue() / (double) totalProducts) * 100);
        }

        return categoryPercentage;
    }
    @Override
    public Double getAveragePrice() {
        return productRepository.calculateAveragePrice();
    }
    @Override
    public Double calculateAverageStockQuantity() {
        return productRepository.getAverageStockQuantity();
    }
    @Override
    public Whishlist addProductToWishlist(Long wishlistId, Long productId) {
        Whishlist wishlist = whishlistRepository.findById(wishlistId)
                .orElseThrow(() -> new RuntimeException("Wishlist not found"));
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found"));

        if (wishlist.getProducts().contains(product)) {
            throw new RuntimeException("Product already in wishlist");
        }

        wishlist.getProducts().add(product);
        return whishlistRepository.save(wishlist);
    }
    @Override
    public Whishlist removeProductFromWishlist(Long wishlistId, Long productId) {
        Whishlist wishlist = whishlistRepository.findById(wishlistId)
                .orElseThrow(() -> new RuntimeException("Wishlist not found"));
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found"));

        if (!wishlist.getProducts().contains(product)) {
            throw new RuntimeException("Product not in wishlist");
        }

        wishlist.getProducts().remove(product);
        return whishlistRepository.save(wishlist);
    }
    @Override
    public List<Product> getWishlistProducts(Long wishlistId) {
        Whishlist wishlist = whishlistRepository.findById(wishlistId)
                .orElseThrow(() -> new RuntimeException("Wishlist not found"));
        return wishlist.getProducts();
    }
    @Override
    public Review addReview(Long productId, int rating) {
        // Fetch the Product by its id
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found"));

        // Create a new Review and set the Product and rating
        Review review = new Review();
        review.setProduct(product);
        review.setRating(rating);

        // Save the Review to the database
        return reviewRepository.save(review);
    }

    public List<Review> getReviews(Long productId) {
        return reviewRepository.findByProductIdProduct(productId);
    }
    @Override

    // Method to update review
    public Review updateReview(Long reviewId, int rating) {
        Optional<Review> existingReview = reviewRepository.findById(reviewId);
        if (existingReview.isPresent()) {
            Review review = existingReview.get();
            review.setRating(rating);
            return reviewRepository.save(review);
        }
        throw new RuntimeException("Review not found");
    }
    @Override
    public PhoneNumber savePhoneNumber(PhoneNumber phoneNumber) {
        return phoneNumberRepository.save(phoneNumber);
    }
    @Override
    public List<String> getAllNumbers() {
        return phoneNumberRepository.findAll()
                .stream()
                .map(PhoneNumber::getNumber)
                .toList();
    }








}

File: client/sportyfinalbackfront-master/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product } from '../models/product';
import { Review } from '../models/review';
import { Comment } from '../models/Comment.model';

@Injectable({
  providedIn: 'root',
})
export class ProductService {
  private apiUrl = 'http://localhost:8090/products'; // Adjust backend URL

  constructor(private http: HttpClient) {}

  // Get all products
  getAllProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/all`);
  }

  // Get product by ID
  getProductById(id: number): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/${id}`);
  }

  // Add a new product (sending JSON instead of FormData)
  addProduct(product: Product): Observable<Product> {
    return this.http.post<Product>(`${this.apiUrl}/add`, product);
  }

  // Update an existing product (sending JSON instead of FormData)
  updateProduct(id: number, product: Product): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/update/${id}`, product);
  }

  // Delete a product
  deleteProduct(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/delete/${id}`);
  }
  assignProductToShop(productId: number, shopId: number): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/${productId}/assign-shop/${shopId}`, {});
  }
  getProductsByPriceRange(minPrice: number, maxPrice: number): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/filterByPrice?minPrice=${minPrice}&maxPrice=${maxPrice}`);
  }
  searchProductsByName(productName: string): Observable<Product[]> {
    return this.http.get<Product[]>(`${this.apiUrl}/search?name=${productName}`);
  }
  getCategoryDistribution(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/category-distribution`);
  }
    // Get average price of all products
    getAveragePrice(): Observable<number> {
      return this.http.get<number>(`${this.apiUrl}/average-price`);
    }
    getAverageStockQuantity(): Observable<number> {
      return this.http.get<number>(`${this.apiUrl}/average-stock`);
    }
    addProductToWishlist(wishlistId: number, productId: number): Observable<any> {
      return this.http.post<any>(`${this.apiUrl}/${wishlistId}/add-to-wishlist/${productId}`, {});
    }
  
    // Remove product from wishlist
    removeProductFromWishlist(wishlistId: number, productId: number): Observable<any> {
      return this.http.delete<any>(`${this.apiUrl}/${wishlistId}/remove-from-wishlist/${productId}`);
    }
  
    // Get all products in a wishlist
    getWishlistProducts(wishlistId: number): Observable<Product[]> {
      return this.http.get<Product[]>(`${this.apiUrl}/${wishlistId}/wishlist-products`);
    }
    // Add review to a product
addReview(productId: number, rating: number): Observable<Review> {
  return this.http.post<Review>(`${this.apiUrl}/${productId}/reviews?rating=${rating}`, null);
}

// Update an existing review
updateReview(reviewId: number, rating: number): Observable<Review> {
  return this.http.put<Review>(`${this.apiUrl}/reviews/${reviewId}`, { rating });
}

// Get all reviews for a product
getReviews(productId: number): Observable<Review[]> {
  return this.http.get<Review[]>(`${this.apiUrl}/${productId}/reviews`);
}
getCommentsByProduct(productId: number): Observable<Comment[]> {
  return this.http.get<Comment[]>(`${this.apiUrl}/product/${productId}`);
}

addComment(productId: number, comment: Comment): Observable<Comment> {
  return this.http.post<Comment>(`${this.apiUrl}/${productId}/comments`, comment);
}

// Delete a comment by ID (you may want to implement this in the backend)
deleteComment(id: number): Observable<void> {
  return this.http.delete<void>(`${this.apiUrl}/comments/${id}`);
}

// Like a comment
likeComment(id: number): Observable<Comment> {
  return this.http.put<Comment>(`${this.apiUrl}/comments/${id}/like`, {});
}

// Dislike a comment
dislikeComment(id: number): Observable<Comment> {
  return this.http.put<Comment>(`${this.apiUrl}/comments/${id}/dislike`, {});
}
// Get recommended products based on age and budget
getRecommendedProducts(age: number, budget: number): Observable<any[]> {
  return this.http.get<any[]>(`${this.apiUrl}/recommend?age=${age}&budget=${budget}`);
}


    
  
 
}

  

